### ISSUE TITLE
Is there a difference between running by GPU and CPU?

### ISSUE THREAD

**Issue by jiaying002** [geoschem/geos-chem#1767]
Name: Jiaying
Institution: Nanyang Technological University

### Description of your issue or question
My current running time to run a one-year global 4*5 resolution simulation took me a month. I found there are people who only need 5 hours. I originally used 1 node -16 cores (2476 Intel Skylake Xeon(R) Gold 6142 with 384GB RAM), then I switched to 1 node-20 GPGPU (Intel Skylake Xeon(R) Gold 6150 with 768GB RAM), and the time for simulation didn't change much.

Hence, I'd like to know 
1) is there a difference in running time between running by GPU and CPU?
2) if the code needs to be changed when switching to GPU?
3) I found the XGBoost seems pretty useful to speed up the simulation time. I'd like to know if there is a way to use it. https://ntrs.nasa.gov/citations/20190033152
4) Would you please help me check which one is the best to run GEOS-Chem, which could result in a shorter simulation time?
![image](https://user-images.githubusercontent.com/113654537/233527079-1af7d472-d82c-4fc8-a301-1da901c63867.png)

Thank you in advance!

**Comment by yantosca** [comment]
Thanks for writing @jiaying002.  We do not support GEOS-Chem on GPUs.  I believe to take advantage of GPUs, you would need to use the LLVM compilers (i.e. clang and flang) instead of the GNU compilers (gcc and gfortran), and GEOS-Chem has not been modified accordingly yet.  

Also, many GPUs do not support double-precision math, but instead use "bfloat" precision, which is a 16-bit datatype.   We have shown that the full-chemistry simulations in GEOS-Chem become unstable unless you use 64-bit precision (i.e. REAL*8), because the reaction rates use lots of EXP, log, and raising numbers to powers.  This means that you need the 64-bit precision in order to avoid underflow (since a 32-bit number, aka REAL*4 can only represent values in the range -1e-38 to 1e-38.)

It would be an interesting research project to see if GEOS-Chem could be made compatible with GPUs but this is beyond the scope of what the GEOS-Chem Support Team could do.

**Comment by lizziel** [comment]
Hi @jiaying002, adding on to what @yantosca said, I am wondering what simulation you are doing that takes a full month for a global 1-year run. Our 4x5 1-year benchmark simulations take about a week. It is also surprising that some runs can go in 5 hrs. Could you give more information on the simulation type you are using, the simulation type used by the people who can run it in 5 hrs, and how many processors you/they are using? Also, are you using any special compiler options, such as running in debug mode? That can slow down the model.

For the list of processors, you could try doing a very short run using each, assuming they are simple to request, and compare the timing.

**Comment by yantosca** [comment]
@jiaying002 I recently ran a couple of GC-Classic 14.2.0 (in development) 1-month fullchem runs with gfortran 10 and gfortran 12 on 48 Intel Cascade-Lake CPUs.  The runs took about 5.5 hours.

I wonder if you are not setting OMP_NUM_THREADS properly in your environment.  As @lizziel mentioned, you may be only running on a single CPU.

**Comment by jiaying002** [comment]
Thank you so much for helping me!!! @yantosca @lizziel 

@yantosca Thank you for telling me that GPU cannot work on GEOS-Chem. You are correct, I did set the OMP_NUM_THREADS to 1, and I didn't notice that. Now, after I set the core number to 32, the simulation could be finished within 37 hours! Thank you so much!

@lizziel Thank you so much for asking! After further discussion, it turned out that the 5 hours run only simulated the CO2, not full-chem. So, that's the reason why the simulation finished that fast.
